### ISSUE TITLE
Mixin Tweaker

### ISSUE THREAD

**Issue by LoganDark** [DemonWav/MinecraftDevIntelliJ#653]
So, I know that MinecraftDev supports creating Forge projects, Liteloader projects, whatever... But what about just regular old Mixin tweakers?

For example, Baritone is a Mixin tweaker. It uses ForgeGradle, but without Forge, so it's just an MCP project. It uses Mixin to modify Minecraft, and it even gets compiled to a jar where Forge can call the tweaker as a coremod.

MinecraftDev has no issue dealing with Mixin tweakers. In fact, it brings a few blessings to the table that I can't live without (mixin debugging!), but it still doesn't have a proper project _template_, you know, a checkbox in the create project dialog.

https://github.com/cabaletta/baritone/blob/master/build.gradle

I looked at #626 for one second and saw that there was no detail, then it was marked as stale and closed :( so I decided to make my own.

**Comment by DenWav** [comment]
I'm not too sold on this idea. The templates are just there to be quick and helpful, they don't aim to cover 100% of all project types. I would consider a bare tweaker as a more advanced type of project that a template wouldn't be as useful for.

**Comment by LoganDark** [comment]
> I'm not too sold on this idea. The templates are just there to be quick and helpful, they don't aim to cover 100% of all project types.

True, but projects are also useful for avoiding trying to set up your fucking build for days, which is what I've been trying to deal with. I've gotten it to the point where it outputs a Forge jar that Forge loads and runs, but the refMap is screwed up and doesn't include field names.

> I would consider a bare tweaker as a more advanced type of project that a template wouldn't be as useful for.

True, but I think MinecraftDev needs an MCP project and this seems perfect. Ever since I started working on Baritone I've been in love with MCP and not having to deal with Forge's shit. For client side-only mods, like minimaps or HUDs or whatever, MCP is perfect. Maybe even for custom clients that you can still also use with Forge.

The dev experience is so much better because you don't have to wait an hour for Forge to load every time. It's just 5 seconds. You don't have to spend time looking for Forge hooks. You can add your own. You don't need to manipulate bytecode, Mixin does it for you.

Since Mixin does not have a guide for setting it up, ForgeGradle does not have a guide on using the MCP, and gradle can be complicated trying to get this working, a starter project would be _extremely_ helpful.

I'd be glad to provide the template, using stupid regular groovy build.gradle because Kotlin is just causing all sorts of issues :(

**Comment by LoganDark** [comment]
a

**Comment by LoganDark** [comment]
a

**Comment by DenWav** [comment]
I think the answer on this is no... @Mumfrey thoughts?

**Comment by LoganDark** [comment]
Doesn't have to be exhaustive, after all you're always meant to configure things to suit your use case, but simply a _template_ to make the process of setting up all that tedious boilerplate a _little_ easier would be _much_ appreciated.

**Comment by Mumfrey** [comment]
> I think the answer on this is no... @Mumfrey thoughts?

I mean I don't know what I can add, you kind of already said what needed to be said up-front and it's your project:

> I'm not too sold on this idea. The templates are just there to be quick and helpful, they don't aim to cover 100% of all project types. I would consider a bare tweaker as a more advanced type of project that a template wouldn't be as useful for.

Mixins standalone (eg. not hooked into any particular ecosystem, just running raw inside a transformer framework) is not really a template-able thing for a number of reasons, and honestly shouldn't be. If you want complete control of your environment then *by definition* you're not going to want to use a template. Plus defining the stuff in gradle is super straightforward as long as you undertstand the piplelines in question, and if you **don't** understand them then you probably shouldn't be doing a custom environment anyway, so it's kind of self-fulfilling.

*ModLauncher* as an environment is kind of better, because the contracts are better defined and spinning up a "vanilla" modlauncher is pretty easy. But LaunchWrapper doesn't really have a no-op tweaker because at the very least tweakers have to handle program arguments correctly, and the mixin tweaker doesn't do that so a custom primary tweaker is needed anyway.

> Since Mixin does not have a guide for setting it up, ForgeGradle does not have a guide on using the MCP, and gradle can be complicated trying to get this working, a starter project would be _extremely_ helpful.

This kind of reinforces the point though. Reading and understanding ForgeGradle's code for example, is an important first step in setting up a custom environment (it's actually mind-blowingly simple to set up a "vanilla" workspace in FG 2.3 *and* 3.0 provided you spend 10 minutes reading the source) and it's pretty much deliberately not documented for this precise reason.

tl;dr: no this shouldn't be templated because it's already simple if you know what you're doing, and if you *don't* know what you're doing you shouldn't be doing it anyway.
